### ISSUE TITLE
Performance issues parsing large networks

### ISSUE THREAD

**Issue by marvinvanaalst** [bioasp/meneco#29]
Versions used:
```
clyngor    | 0.3.31
Meneco     | 2.0.1
```
---
This is probably more a clyngor issue, but I'll post it here because that's where it affected me :)

I noticed a heavy performance drop when trying to work with large models, so while a ~4207 reactions E. coli sbml file parsed in roughly 7 seconds, a 27742 reactions metacyc sbml file (both including certain duplications) parsed in 10 minutes! In comparison, both together took roughly 5 seconds on meneco 1.5.3

I used the smaller one to profile on where that performance regression comes from

```python
import cProfile
cProfile.run('readSBMLnetwork("ecoli.sbml", "draft")')
```
and noticed the following oddity (I'll spare you the rest of the output)
```
160214 function calls in 7.715 seconds

Ordered by: standard name

ncalls  tottime  percall  cumtime  percall filename:lineno(function)
...
24724    6.163    0.000    6.178    0.000 as_pyasp.py:78(add)
....
```
So a whopping 6 of 7 seconds is used on the `TermSet.add` method. Thus I quickly ran a check on how this function scales over more input, and it seems to be roughly exponential.

![output](https://user-images.githubusercontent.com/31540811/165305909-c0c793c3-0ac1-45d6-ac3d-88b93092842c.png)

I don't know exactly why they used `frozenset`s in there, but that line here will create two new ones every time the function is called, which probably explains the runtime.

```python
def add(self, atom: Atom):
    self._terms |= frozenset({atom})
```

I don't know if you can work around that (I haven't checked your source yet) or if the clyngor folks would be open to change, but in the meantime, I will have to downgrade our [moped](https://gitlab.com/marvin.vanaalst/moped) software, because some of our users ran into problems due to this.

Hope we can quickly find a fix for this, if I can help just let me know, I'd be happy to.

Cheers

**Comment by ArnaudBelcour** [comment]
Hi,

I am not a developer of meneco but I have work on the implementation of clyngor in [miscoto](https://github.com/cfrioux/miscoto) and [menetools](https://github.com/cfrioux/MeneTools) and I have not encountered this issue in these implementations.

I have compared the implementation reading the SBML files in [meneco](https://github.com/bioasp/meneco/blob/master/meneco/sbml.py#L112-L168) and [miscoto](https://github.com/cfrioux/miscoto/blob/master/miscoto/sbml.py#L200-L266). To test it I use a SBML file containing most reactions of MetaCyc. Indeed the function of meneco takes several minutes whereas the function of miscoto reads the input in a few seconds.

The main difference between the two is that instead of using a [TermSet](https://github.com/bioasp/meneco/blob/master/meneco/sbml.py#L114) and then add Atoms in it, miscoto uses a simple [set](https://github.com/cfrioux/miscoto/blob/master/miscoto/sbml.py#L210) than convert it into a [TermSet](https://github.com/cfrioux/miscoto/blob/master/miscoto/sbml.py#L264) after adding Atoms in it.

I replace the TermSet by a set in meneco and apply the conversion into a TermSet after adding all the atoms and it seems to fix the issue.

I am not an expert in python but maybe this is linked to the fact that the `add` function of TermSet will make an union each time something is added compared to the `add` function of the set.

So a fix could be to replace the TermSet by a set and after adding the different Atoms, convert it into a TermSet.

Best Regards,
Arnaud Belcour.

**Comment by sthiele** [comment]
@marvinvanaalst  Thanks for reporting the issue.
I applied the changes proposed by @ArnaudBelcour in branch [29-performance-issues-parsing-large-networks](https://github.com/bioasp/meneco/tree/29-performance-issues-parsing-large-networks). It looks good so far .. while I'm still running tests you may also want to test it before I create a new release

**Comment by marvinvanaalst** [comment]
Parsing is down to ~7s again with that branch. Looking good!

**Comment by sthiele** [comment]
I just released version 2.0.2 with the fix.

**Comment by marvinvanaalst** [comment]
And successfully integrated into [moped](https://pypi.org/project/moped/). Thanks for the quick work! :)
